skbio/tree/_tree.py

```diff
@@ -3116,6 +3119,54 @@ def assign_supports(self):
             else:
                 node.support, node.name = node._extract_support()
 
+    def is_bifurcating(self, strict=False):
+        r"""Check if the tree is bifurcating.
+
+        .. versionadded:: 0.6.3
+
+        Parameters
+        ----------
+        strict : bool, optional
+            Whether to consider single-child nodes as violations of bifurcation.
+            Default is False.
+
+        See Also
+        --------
+        bifurcate
+        prune
+
+        Notes
+        -----
+        In a bifurcating tree (a.k.a. binary tree), every node has at most two
+        children. The property of bifurcation is necessary for a wide range of tree
+        analyses. In contrast, if a node has three or more children, it is considered
+        as multifurcating, or polytomy in phylogenetics.
+
+        In strict mode, every internal node (including root) has to have exactly two
+        children in order for the tree to be bifurcating. Single-child nodes are
+        considered as violations. These nodes can be collapsed by :meth:`prune`.
+
+        This method operates on the subtree below the current node.
+
+        Examples
+        --------
+        >>> from skbio import TreeNode
+        >>> tree = TreeNode.read(["((a,b,c),(d,e))root;"])
+        >>> tree.is_bifurcating()
+        False
+
+        """
+        if strict:
+            for node in self.non_tips(include_self=True):
+                if len(node.children) != 2:
+                    return False
+            return True
+        else:
+            for node in self.traverse(include_self=True):
+                if len(node.children) > 2:
+                    return False
+            return True
```

Comment:
A redundant return can be removed with:

```suggestion
        else:
            for node in self.traverse(include_self=True):
                if len(node.children) > 2:
                    return False
        return True
```

A further reduction in replication could be done with the below, though for something like this I think it's really just a stylistic item. 

```python
import operator

if strict:
    test = operator.ne
else:
    test = operator.gt

for node in self.non_tips(include_self=True):
    if test(len(node.children), 2):
        return False

return True
```

Reply:
@wasade Interesting! Your suggestion is very elegant. I think that `test` as a function call may be slower than a straight operator, but perhaps Python has optimization for this. Let me benchmark it.

Reply:
Possibly though this isn't a performance critical method unless there is a reasonable need to run `is_bifurcating` a very large number of times on the same tree?